DiamondLightSource/blueapi
# Speculative visit directory service

*PLEASE DO NOT MERGE!*
This PR is intended as a speculative prototyping exercise only.

**Design**

In order to comply with the DLS visit directory structure, we need to group acquisitions into files. There are many solutions to this, the basics of this one:

![image](https://github.com/DiamondLightSource/blueapi/assets/29771545/e4117d2b-18af-4538-b3f5-fec1f682cbff)

1. A new service is responsible for determining where data will go for a particular acquisition. This functionality is separated from all other applications for two reasons:
    * So that other data collection applications can also use it each one is guaranteed unique data locations. 
    * This involves making directories and therefore requires write access to places where data goes. For security reasons, only trusted code should have that level of privilege, and it should be kept small and isolated.
3. Every time a group of devices is staged, blueapi will call out to this service (or use a dummy version for offline development) and get a unique data directory for this stage. This is achieved via a plan preprocessor that detects groups of stages and makes the call. 
4. Compatible Ophyd detectors will be aware of a mutable state which is updated when this call occurs, so they can write to the "current" visit.
5. The current group is bundled into the run start metadata. Allowing for a group:run mapping of 1:1 or 1:many. The use case for the latter is writing many runs to a single file for performance reasons.
6. Any aggregating service, such as a nexus file writer, can take the group information from the run starts and use it how it likes. Our initial thinking was to map each group to a nexus file and reach run to an entry, but that is TBD.

**Design Issues**

I remain _very_ sceptical about point 2! It introduces a shared mutable state between all detectors just to get around the fact that there is no direct, universal way to pass the current visit down to them, by design! The other alternative I explored is to put all support for configuring individual detectors in one place, in blueapi, and if your detector isn't supported it will just write as it normally does. Simple example

```python
def configure_data_writing(
    devices: Iterable[Device],
    collection: DataCollection,
) -> None:
    for device in devices:
        if isinstance(device, FileStoreBase):
            path_template = str(collection.raw_data_files_root)

            # Configure Ophyd Device to setup HDF5 writer
            device.reg_root = "/"
            device.read_path_template = path_template
            device.write_path_template = path_template
```

**What's Prototyped Here?**

The actual contents of this PR:

* A very simple visit service that can create and inform clients about unique directories for "acquisition groups".
* A plan preprocessor that calls out to that service (or use an internal dummy version) to create a new group when a set of devices is staged. It also puts metadata into the run start document to show which group it is a part of. Each start document now contains a key called `data_collection_number`, which can be passed back to the visit service for more information. 
* A dummy device that has access to the cached result of the service call.
* A series of tests ensuring that the correct mapping of run -> file is achieved for various use cases, such as multi-run plans.
* A small amendment to the bluesky context that makes blueapi use this preprocessor on every plan it runs (if the plan does not contain any `stage`s, it will be unaffected)

**How to test**

Checkout this branch. To test with the dummy detector, run

```shell
tox -e pytest -- tests/plugins/test_data_writing.py
```

To start the visit service:

```shell
python -m blueapi.plugins.data_writing_server
```

It should run on port 8089, you can test it with

```shell
curl -X POST "http://localhost:8089/collection/mybeamline"
```

Which should create a new group and return a blob of JSON about it.

Running a scan with blueapi should now produce a start document that contains an attribute called `data_collection_number` which we can use to aggregate runs.

**Still to do**

This PR contains a dummy detector with the correct hooks. It would be good to get this working with an actual Ophyd v2 detector, and ideally a v1 one too. We could then run a full collection via blueapi that uses this functionality @coretl @callumforrester  

More discussion on point 2. of the design would be helpful, I think. @coretl @DiamondJoseph @keithralphs @callumforrester  

We should try to get the nexus writer to interpret the group number in the run start document and ensure there are no problems there @DiamondJoseph 

Possibly leverage the `data_session` metadata in the run start schema. @callumforrester  @DiamondJoseph
